core/fragment.c

```diff
@@ -1139,6 +1139,18 @@ hashtable_fragment_reset(dcontext_t *dcontext, fragment_table_t *table)
         if (!dynamo_exited && !dynamo_resetting)
             ASSERT_TABLE_SYNCHRONIZED(table, WRITE);
     });
+#    if !defined(DEBUG) && defined(CLIENT_INTERFACE)
+    if (!dr_fragment_deleted_hook_exists())
+        return;
+    /* i#4226: Avoid the slow deletion code and just invoke the event. */
+    for (i = 0; i < table->capacity; i++) {
+        f = table->table[i];
+        if (!REAL_FRAGMENT(f))
+            continue;
+        instrument_fragment_deleted(dcontext, f->tag, f->flags);
```

Comment:
Nit: Can you explain how does DR know that all fragments can be correctly passed to  instrument_fragment_deleted please? When I looked at how instrument_fragment_deleted is used, particularly in fragment_delete, I see it is conditioned:

```
#ifdef CLIENT_INTERFACE
    if (dr_fragment_deleted_hook_exists() &&
        (!TEST(FRAGDEL_NO_HEAP, actions) || !TEST(FRAGDEL_NO_FCACHE, actions)))
        instrument_fragment_deleted(dcontext, f->tag, f->flags);
#endif
```

Reply:
It's this function that decides those `actions` flags, and the fragment_delete() call a few lines below is not passing either of those NO_HEAP or NO_FCACHE flags, so we know the condition is always true for fragment deletion via hashtable_fragment_reset().  Those NO_ calls are particular types of cleanup done at other times, not at reset/exit.  I can add a comment.

Reply:
Ah I see, yes I agree.

Reply:
I guess the concern would be some future mismatch in conditions between what are now two separate calls to instrument_fragment_deleted.  I can add a comment to the other call as well to keep them in synch just so future devs are aware of the two calls.